### DISCUSSION TITLE
Shared albums across instances

### DISCUSSION THREAD

**Discussion by MrModest** [immich-app/immich#13620] [Feature Request]
Sharing albums within one instance makes sense for a small family living in the same house. But what if I want to share an album with my friend who has his own instance of self-hosted immich?

At some level, I guess, you can call it a Fediverse feature.

Let's say we set up a Tailscale/VPN/etc. between our home servers. So reaching another server via IP isn't a problem. 

I think, just linking between isn't reliable approach in my opinion, so I see it as each instance has its own copy of all media files. 

But mixing my personal photos and shared one in one folder I don't want either.

So, here's how I see this feature.

Before being able to share photos between users from different instances, those instances should be "paired". Essentially somewhere in the DB should be a table that stores metadata about foreign instances and foreign users and also connection info (IP/domain/etc). 

Once pairing is done, sharing should with a user of foreign instances should look like with an internal user. And an album can be shared at the same time with multiple internal and foreign users.

Along with the internal library (in `upload/library/user_id`), immich can have a dedicated folders for media files copied from another instances (for example, `upload/foreign_instances/foreign_instance_ID/foreign_user_ID`).

Inside `foreigh_user_ID`, all folders and media files should be stored in the same way as for other internal users (independently of the way how foreign users library is structured on their own instance). And it contains all medias shared at least in one of the common shared album (so we don't have duplicates even if one photo presents in more that one shared album). Once media files are removed from all shared albums (or no more  internal users participate in the shared albums where these media files are presented), the foreign file can be removed from the internal instance. 

Any (or each) internal user can copy foreign photos into their own library. In this case photos will be copied into their own library folder and remains even if the shared album disappears. 

Potential challenges:
- How to uniquely identify foreign instances if they aren't exposed with a public IP. What if something has changed in the VPN/Tailscale and each server inside network gets another IP. 
  - There should be a way to manually adjust configs of foreign instances, so immich won't recognise it as a new one.
- Uniqueness of IDs (of users, assets and albums) across different instances.
  - immich uses UUID, so theoretically it shouldn't be a big problem.

**Comment by bo0tzz** [comment]
#728
